### PR TITLE
Generators can add source files originating from a StringBuilder

### PR DESCRIPTION
**Pull request by YairHalberstadt**
Fixes https://github.com/dotnet/roslyn/issues/49082

### CODE REVIEW

**Comment by YairHalberstadt** [comment]
I thought about adding the following optimization to StringBuilderText on .NetCore to improve performance of indexing into the builder (which is O(n)): https://gist.github.com/YairHalberstadt/1df8fbeb846d6de0f4b701af314df26a

Do you think it's worth exploring further?

**Comment by 333fred** [comment]
> I thought about adding the following optimization to StringBuilderText on .NetCore to improve performance of indexing into the builder (which is O(n)): [gist.github.com/YairHalberstadt/1df8fbeb846d6de0f4b701af314df26a](https://gist.github.com/YairHalberstadt/1df8fbeb846d6de0f4b701af314df26a)
> 
> Do you think it's worth exploring further?

Possibly. What about it makes it netcore only? And do you have perf numbers on the change?

File: src/Compilers/Core/Portable/Text/StringBuilderText.cs [hunk 1]
```diff
@@ -21,7 +21,7 @@ internal sealed partial class StringBuilderText : SourceText
 
         private readonly Encoding? _encodingOpt;
 
-        public StringBuilderText(StringBuilder builder, Encoding? encodingOpt, SourceHashAlgorithm checksumAlgorithm)
+        public StringBuilderText(StringBuilder builder, Encoding? encodingOpt, SourceHashAlgorithm checksumAlgorithm = SourceHashAlgorithm.Sha1)
```
Comment (333fred) [inline src/Compilers/Core/Portable/Text/StringBuilderText.cs:24]: Why this change?

**Comment by YairHalberstadt** [comment]
> What about it makes it netcore only

It uses the GetChunks method which is .NetCore only: https://docs.microsoft.com/en-us/dotnet/api/system.text.stringbuilder.getchunks?view=net-5.0

File: src/Compilers/Core/Portable/SourceGeneration/GeneratorContexts.cs [hunk 2]
```diff
@@ -70,9 +70,16 @@ internal GeneratorExecutionContext(Compilation compilation, ParseOptions parseOp
...
         /// <param name="hintName">An identifier that can be used to reference this source text, must be unique within this generator</param>
-        /// <param name="source">The source code to be add to the compilation</param>
+        /// <param name="source">The source code to add to the compilation</param>
         public void AddSource(string hintName, string source) => AddSource(hintName, SourceText.From(source, Encoding.UTF8));
 
+        /// <summary>
+        /// Adds source code in the form of a <see cref="StringBuilder"/> to the compilation.
+        /// </summary>
+        /// <param name="hintName">An identifier that can be used to reference this source text, must be unique within this generator</param>
+        /// <param name="source">The source code to add to the compilation. This must not be modified from now on.</param>
```
Comment (333fred) [inline src/Compilers/Core/Portable/SourceGeneration/GeneratorContexts.cs:80]: This is where I _really_ want move semantics... I think a stronger assertion is needed, perhaps "This builder must not be further modified: undefined behavior will result."?

File: src/Compilers/Core/Portable/Text/StringBuilderText.cs (hunk 1 above)
Comment (YairHalberstadt) [reply src/Compilers/Core/Portable/Text/StringBuilderText.cs:24]: We needed to put Sha1 somewhere and this was consistent with the other APIs

**Comment by YairHalberstadt** [comment]
> And do you have perf numbers on the change?

Turns out it's not going to be much better. Whilst the current behavior is quadratic, it takes huge StringBuilders for it to be faster (ones made up of over 100,000 separate strings).

File: src/Compilers/Core/Portable/SourceGeneration/GeneratorContexts.cs (hunk 2 above)
Comment (YairHalberstadt) [reply src/Compilers/Core/Portable/SourceGeneration/GeneratorContexts.cs]: I added stronger language and put it in summary as well.

**Comment by 333fred** [comment]
> it takes huge StringBuilders for it to be faster (ones made up of over 100,000 separate strings).

I'm curious, is it the _same_ for smaller numbers, or is about the same? Because unfortunately, if anyone is going to have these, it's going to people using source generators...

File: src/Compilers/Core/Portable/Text/StringBuilderText.cs (hunk 1 above)
Comment (333fred) [reply src/Compilers/Core/Portable/Text/StringBuilderText.cs:24]: I'd prefer to keep the constructor as is and put the algorithm at the usage site. The only other usage of the constructor does explicitly need to pass along the algorithm from its own arguments, and for these scenarios I prefer to keep the knowledge where it's actually applicable.

**Comment by YairHalberstadt** [comment]
For small strings it's about 6 times worse

**Comment by 333fred** [comment]
> For small strings it's about 6 times worse

Then let's leave it alone :)

**Comment by RikkiGibson** [comment]
Maybe a silly question, but can we use a heuristic where the API decides if the stringbuilder is complex enough to avoid realizing it, then it does so, and if it's too simple, it just realizes it and uses the string path?

That said, since the breaking point is fairly large, this still may not be worth it.

File: src/Compilers/Core/Portable/Text/StringBuilderText.cs (hunk 1 above)
Comment (YairHalberstadt) [reply src/Compilers/Core/Portable/Text/StringBuilderText.cs]: Done

**Comment by YairHalberstadt** [comment]
@RikkiGibson the optimization I'm suggesting is the one in this gist: https://gist.github.com/YairHalberstadt/1df8fbeb846d6de0f4b701af314df26a

It's not about avoiding realizing the string, but about reducing the quadratic overhead of StringBuilder indexer.

It's only possible on .Net Core, as on .Net framework the StringBuilder.GetChunks method doesn't exist.

**Comment by RikkiGibson** [comment]
Thanks, sorry for my ignorance :)

File: src/Compilers/Core/Portable/SourceGeneration/GeneratorContexts.cs [hunk 3]
```diff
@@ -70,9 +70,17 @@ internal GeneratorExecutionContext(Compilation compilation, ParseOptions parseOp
...
+        /// <param name="source">The source code to add to the compilation</param>
         public void AddSource(string hintName, string source) => AddSource(hintName, SourceText.From(source, Encoding.UTF8));
 
+        /// <summary>
+        /// Adds source code in the form of a <see cref="StringBuilder"/> to the compilation.
+        /// This builder must not be further modified: undefined behavior will result.
+        /// </summary>
+        /// <param name="hintName">An identifier that can be used to reference this source text, must be unique within this generator</param>
+        /// <param name="source">The source code to add to the compilation. This builder must not be further modified: undefined behavior will result.</param>
+        public void AddSource(string hintName, StringBuilder source) => AddSource(hintName, new StringBuilderText(source, Encoding.UTF8, SourceHashAlgorithm.Sha1));
```
Comment (jaredpar) [inline src/Compilers/Core/Portable/SourceGeneration/GeneratorContexts.cs:82]: We cannot hard code `Sha1` in the code anymore. The only reason we can have `Sha1` at all is for compatibility reasons. Most uses of `Sha` in the code base you see exist for this reason. 

@chsienki short term we need to make sure that all of the `SourceText` we add here use `SourceHashAlgorithm.Sha256` (temporarily strong but sure to be a problem at some point down the road). Long term I think we need to feed through the hash algorithm specified on the command line to the `Compilation` such that files added by generators inherit this.

**Review by jaredpar — DISMISSED** [review]
Cannot add `Sha1` to the code.

File: src/Compilers/Core/Portable/SourceGeneration/GeneratorContexts.cs (hunk 3 above)
Comment (chsienki) [reply src/Compilers/Core/Portable/SourceGeneration/GeneratorContexts.cs]: @jaredpar, makes sense. I'll open an issue to track this I'm the generators code.

Comment (YairHalberstadt) [reply src/Compilers/Core/Portable/SourceGeneration/GeneratorContexts.cs]: I updated 1 of the cases in this PR.

**Review by 333fred — COMMENTED** [review]
Only thing I think is left is a test to cover the API.

**Review by 333fred — APPROVED** [review]
LGTM (commit 6)

**Comment by 333fred** [comment]
@chsienki for a second review.

File: src/Compilers/Core/Portable/SourceGeneration/GeneratorContexts.cs [hunk 4]
```diff
@@ -70,8 +70,16 @@ internal GeneratorExecutionContext(Compilation compilation, ParseOptions parseOp
         /// Adds source code in the form of a <see cref="string"/> to the compilation.
         /// </summary>
         /// <param name="hintName">An identifier that can be used to reference this source text, must be unique within this generator</param>
-        /// <param name="source">The source code to be add to the compilation</param>
-        public void AddSource(string hintName, string source) => AddSource(hintName, SourceText.From(source, Encoding.UTF8));
+        /// <param name="source">The source code to add to the compilation</param>
+        public void AddSource(string hintName, string source) => AddSource(hintName, SourceText.From(source, Encoding.UTF8, SourceHashAlgorithm.Sha256));
```
Comment (chsienki) [inline src/Compilers/Core/Portable/SourceGeneration/GeneratorContexts.cs:74]: >SourceHashAlgorithm.Sha256 [](start = 124, length = 26)

I'd like to think through the implications of changing this slightly more before we do so. I agree we should use 256 below where we're forced to supply an algorithm, but I don't think we should make the change here as part of this PR.

File: src/Compilers/Core/Portable/SourceGeneration/GeneratorContexts.cs [hunk 5]
```diff
@@ -70,8 +70,16 @@ internal GeneratorExecutionContext(Compilation compilation, ParseOptions parseOp
...
+
+        /// <summary>
+        /// Adds source code in the form of a <see cref="StringBuilder"/> to the compilation.
+        /// This builder must not be further modified: undefined behavior will result.
+        /// </summary>
+        /// <param name="hintName">An identifier that can be used to reference this source text, must be unique within this generator</param>
+        /// <param name="source">The source code to add to the compilation. This builder must not be further modified: undefined behavior will result.</param>
```
Comment (chsienki) [inline src/Compilers/Core/Portable/SourceGeneration/GeneratorContexts.cs:81]: >must not be further modified [](start = 89, length = 28)

What happens if I get the `StringBuilder` from a `StringBuilderCache` becuase I'm trying to be more efficient?

**Comment by chsienki** [comment]
So, in general I'm not sure if we want to go this route. The fact that there's no way to 'Freeze' the `StringBuilder` seems like it could cause some hard to track down issues in the IDE if the generator starts changing it from underneath us. 

Right now we don't parse the added files until the end of the execution method, so generator that added it first and then appended it would probably be ok, and the issue would only be if it held onto it during executions passes and modifying it subsequent iterations (though I've no doubt someone would take a dependency on that, and if we ever changed the parsing order we'd break someone).

File: src/Compilers/Core/Portable/SourceGeneration/GeneratorContexts.cs (hunk 5 above)
Comment (jaredpar) [reply src/Compilers/Core/Portable/SourceGeneration/GeneratorContexts.cs:81]: Yeah this seems like a particularly dangerous problem here that I would want to guard against.

Comment (jaredpar) [reply src/Compilers/Core/Portable/SourceGeneration/GeneratorContexts.cs:81]: In general the policy is to not pass around mutable state between the compiler and plugins that would impact the output of the build. I hadn't caught on in first review that we don't immediately realize the contents of the `StringBuilder` here.

**Comment by jcouv** [comment]
It's unfortunate, but I'll go ahead and close this PR following discussion with Chris. Thanks for the proposed change

**Comment by alrz** [comment]
Could we accept another type like TextReader? I've seen APIs accepting that in addition to a string in case you want to stream instead of allocating the whole thing in a string.

**Comment by jcouv** [comment]
I'll let @chsienki comment on that (can we avoid allocating the string?)

**Comment by jaredpar** [comment]
I don't think we can. The main point @chsienki is making is that `Compilation` objects are immutable. Every source type we accept from generators must be immutable else we end up violating that core principal of `Compilation` objects. The `TextReader` object, like `StringBuilder`, is not immutable. 

This is a core principal of Roslyn that can't really compromise on.

**Comment by alrz** [comment]
I'm surprised that there's no type in BCL that supports this scenario.

**Comment by chsienki** [comment]
Yeah, it's pretty unfortunate. We could potentially ship a `FreezableStringBuilder` that wraps a `StringBuilder` with the same interface, but throws once you 'Freeze' it (by passing it to the compilation). But theres obviously a maintance burden on such a type, so I think we'd have to be sure it's worth it beforehand.

**Comment by jcouv** [comment]
I'd like to clarify the requirements for immutability. I assume that `TextReader` only has APIs to *read*, much like a potential `FreezableStringBuilder`. But I assume there is always risk that the underlying data structure would be changed (so if we read more than once then we could get different results). Is that the concern?

**Comment by alrz** [comment]
Actually if you plan to use System.IO.Pipelines at some point, `ReadOnlySequence<T>` is another type that fits the bill. Though currently there's not many APIs to make it convenient. I tried to construct one manually from `StringBuilder.GetChunks` but it still doesn't guarantee that the data doesn't change underneath unless we only read once in which case that issue is moot.

**Comment by jnm2** [comment]
What about a SourceTextBuilder type in Roslyn that would be freezable and would also be free to manage its own backing storage if that had benefits over an internal StringBuilder?

**Comment by Neme12** [comment]
Doesn't the mutability problem already exist today even without a StringBuilder overload because you can create a SourceText from a TextReader?
https://github.com/dotnet/roslyn/pull/47252/files#r480506021

**Comment by Neme12** [comment]
Actually never mind, it seems like it doesn't, because it reads the TextReader and if it's large, it eventually calls `LargeText.Decode`. Couldn't this implementation create a TextReader over the StringBuilder and call the TextReader to avoid the mutability problem, like @jasonmalinowski  did in the PR I linked above? Also, it seems like instead of adding an overload to AddSource, maybe there should be a SourceText.From overload that takes a StringBuilder added instead.

**Comment by jasonmalinowski** [comment]
Tagging @chsienki on the suggestion -- having a helper SourceText.From would still make this a bit easier to do.

**Comment by Neme12** [comment]
> I don't think we can. The main point @chsienki is making is that Compilation objects are immutable. Every source type we accept from generators must be immutable else we end up violating that core principal of Compilation objects. The TextReader object, like StringBuilder, is not immutable.

> So, in general I'm not sure if we want to go this route. The fact that there's no way to 'Freeze' the StringBuilder seems like it could cause some hard to track down issues in the IDE if the generator starts changing it from underneath us.

@jaredpar @chsienki Since `SourceText` can be derived from by users, users can provide their own implementation that wraps a `StringBuilder` or something else that is mutable. This means that `SourceText` is *not* guaranteed to be immutable. So if a mutable `SourceText` would cause issues in the IDE, isn't that a bug already that should be fixed?

**Comment by jaredpar** [comment]
> This means that SourceText is not guaranteed to be immutable. So if a mutable SourceText would cause issues in the IDE, isn't that an existing bug that should be fixed?

There are many ways customers can break us by implementing APIs that are intented to be immutable in a mutating fashion. Consider that our source generator and analyzer stories have an implicit assumption that customers are abiding by our rules on immutability. Specifically that analyzers should be state free, don't cache `Compilation`, etc ... There is no way for us to enforce this, all we can do is document it and push on customers we find violating our rules. 

At the same time we want to make it as hard as possible for customers to break our abstractions. Taking a dependency on `StringBuilder` not being mutated after a specific point feels a bit dicey to me.

**Comment by Neme12** [comment]
Makes sense. It just seems odd to me that `SourceText` is derivable if it's really supposed to be immutable and the compiler depends on that.
